rln/src/merkle_tree.rs

```diff
@@ -1,251 +1,231 @@
-        let levels = tree_height + 1;
-        // Compute empty node values, leaf to root
-        let empty = successors(Some(initial_leaf), |prev| Some(H::hash_node(prev, prev)))
-            .take(levels)
-            .collect::<Vec<_>>();
-
-        // Compute node values
-        let nodes = empty
-            .iter()
-            .rev()
-            .enumerate()
-            .flat_map(|(levels, hash)| repeat(hash).take(1 << levels))
-            .cloned()
-            .collect::<Vec<_>>();
-        debug_assert!(nodes.len() == (1 << levels) - 1);
-
-        let next_index = 0;
-
-        Self {
-            levels,
-            empty,
-            nodes,
-            next_index,
-        }
+    pub fn root(&self) -> H::Fr {
+        self.get_node(0, 0)
     }
 
-    #[must_use]
-    pub fn num_leaves(&self) -> usize {
-        self.levels
-            .checked_sub(1)
-            .map(|n| 1 << n)
-            .unwrap_or_default()
+    pub fn set(&mut self, index: usize, leaf: H::Fr) -> io::Result<()> {
+        if index >= self.tree_size() {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                "index exceeds set size",
+            ));
+        }
+        self.nodes.insert((self.depth, index), leaf);
+        self.recalculate_from(index);
+        self.next_index = max(self.next_index, index + 1);
+        Ok(())
     }
 
-    #[must_use]
-    pub fn root(&self) -> H::Hash {
-        self.nodes[0]
+    pub fn update_next(&mut self, leaf: H::Fr) -> io::Result<()> {
+        self.set(self.next_index, leaf)?;
+        Ok(())
     }
 
-    pub fn set(&mut self, leaf: usize, hash: H::Hash) {
-        self.set_range(leaf, once(hash));
-        self.next_index = max(self.next_index, leaf + 1);
+    pub fn delete(&mut self, index: usize) -> io::Result<()> {
+        let leaf = H::default_leaf();
+        self.set(index, leaf)?;
+        Ok(())
     }
 
-    pub fn set_range<I: IntoIterator<Item = H::Hash>>(&mut self, start: usize, hashes: I) {
-        let index = self.num_leaves() + start - 1;
-        let mut count = 0;
-        // TODO: Error/panic when hashes is longer than available leafs
-        for (leaf, hash) in self.nodes[index..].iter_mut().zip(hashes) {
-            *leaf = hash;
-            count += 1;
+    pub fn proof(&self, index: usize) -> io::Result<MerkleProof<H>> {
+        if index >= self.tree_size() {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                "index exceeds set size",
+            ));
         }
-        if count != 0 {
-            self.update_nodes(index, index + (count - 1));
-            self.next_index = max(self.next_index, start + count);
+        let mut witness = Vec::<(H::Fr, u8)>::with_capacity(self.depth);
+        let mut i = index;
+        let mut depth = self.depth;
+        loop {
+            i ^= 1;
+            witness.push((self.get_node(depth, i), (1 - (i & 1)).try_into().unwrap()));
```

Comment:
`1` means right and `0` means left  but this computation `(1 - (i & 1))` does the reverse.  Below suggestion (which correspond to the killic lib) works if you use `bool` instead of `u8`. 
```suggestion
            witness.push((self.get_node(depth, i), ((i & 1 == 1)).try_into().unwrap()));
```

Reply:
Fine with adjusting it based on needs, but then this new indexing differs from the conventional indexing (where branches to the left are indexed by 0 and to the right by one). If we could stick to the conventional one, it makes the code more readable and compliant with what is commonly known in the literature. 
BTW, no rush to do it in this PR, another PR would be fine.

Reply:
> where branches to the left are indexed by 0 and to the right by one

Agree this is more natural. I don't quite understand the argument for not doing this?